testlapack/dtrevc3.go

```diff
@@ -0,0 +1,210 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/floats"
+	"github.com/gonum/lapack"
+)
+
+type Dtrevc3er interface {
+	Dtrevc3(side lapack.EigVecSide, howmny lapack.HowMany, selected []bool, n int, t []float64, ldt int, vl []float64, ldvl int, vr []float64, ldvr int, mm int, work []float64, lwork int) int
+}
+
+func Dtrevc3Test(t *testing.T, impl Dtrevc3er) {
+	rnd := rand.New(rand.NewSource(1))
+	for _, side := range []lapack.EigVecSide{lapack.RightEigVec, lapack.LeftEigVec, lapack.RightLeftEigVec} {
+		for _, howmny := range []lapack.HowMany{lapack.AllEigVec, lapack.AllEigVecMulQ, lapack.SelectedEigVec} {
+			for _, n := range []int{0, 1, 2, 3, 4, 5, 10, 34, 100} {
+				for _, extra := range []int{0, 11} {
+					for _, optwork := range []bool{true, false} {
+						for cas := 0; cas < 10; cas++ {
+							tmat := randomSchurCanonical(n, n+extra, rnd)
+							testDtrevc3(t, impl, side, howmny, tmat, optwork, rnd)
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func testDtrevc3(t *testing.T, impl Dtrevc3er, side lapack.EigVecSide, howmny lapack.HowMany, tmat blas64.General, optwork bool, rnd *rand.Rand) {
+	const tol = 1e-14
+
+	n := tmat.Rows
+	extra := tmat.Stride - tmat.Cols
+	right := side != lapack.LeftEigVec
+	left := side != lapack.RightEigVec
+
+	var selected, selectedWant []bool
+	var mWant int
+	if howmny == lapack.SelectedEigVec {
```

Comment:
Added some comments that hopefully clarify it. The loop just mimics what Dtrevc3 does: it takes `selected`, standardizes it based on whether the eigenvalues of `T` are real or complex, and computes how many columns are necessary for storing the eigenvectors. For real eigenvalues it just sets the corresponding element of `selectedWant` to true. For complex eigenvalues, it sets to true only the element corresponding to the first column of the pair.